Start making clippy easier to invoke in non-cargo contexts

jsgf opened the pull request:
Clippy (clippy-driver) currently has a couple of strong but unnecessary couplings with cargo. This series:
1. makes detection of check builds more robust, and
2. make clippy-driver use the --sysroot specified on the command line as its internal sysroot.

jsgf commented:
@oli-obk Clippy's advice for https://travis-ci.com/rust-lang/rust-clippy/jobs/170625109#L1309 is wrong and would not compile - this is an explicit while because it also advances the iterator within the body. Should I just suppress that particular clippy warning for the function?

Manishearth commented:
Yes, please do. And file a bug for that false positive.

phansch commented:
retriggering CI

phansch commented:
Looks like it just needs a small rustfmt fix: https://travis-ci.com/rust-lang/rust-clippy/jobs/173272226#L1397

jsgf commented:
I fixed the formatting, but the tests seem to be failing for unrelated reasons which is preventing me from making sure the tests I added are actually working.

phansch commented:
huh that's a weird CI failure indeed. Some UI tests seem to have no stderr output?

I will trigger a new build and see if it still happens.

bors commented:
:umbrella: The latest upstream changes (presumably #2857) made this pull request unmergeable. Please resolve the merge conflicts.

phansch commented:
The weird UI test failures seem to have gone away now. I'm not sure where they came from, tbh.

It looks like the output of the `cstring.stderr` is different now: https://travis-ci.com/rust-lang/rust-clippy/jobs/174594861#L1040

jsgf commented:
@phansch It hasn't changed, but clippy is being invoked differently I think. I'm still trying to understand how all the ui tests actually get invoked.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit f0131fbab6331eb44d88c94783be2ac9b87f8b06 has been approved by `oli-obk`

<!-- @bors r=oli-obk f0131fbab6331eb44d88c94783be2ac9b87f8b06 -->

bors commented:
:hourglass: Testing commit f0131fbab6331eb44d88c94783be2ac9b87f8b06 with merge e176324fc593644a3ce00aa2ec7c1887d571003e...

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust-clippy/builds/99949464), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/builds/22171136)
Approved by: oli-obk
Pushing e176324fc593644a3ce00aa2ec7c1887d571003e to master...
